### PR TITLE
Replace archive functionality with libarchive

### PR DESCRIPTION
**Pull request by Jubilee101**
I verified the correctness and this seems alright. Server starts normally. Note that this replacement doesn't do anything special with symlinks or tablespaces currently. I think this part should be first handled in restore. So it only does what our archive functionality used to do, nothing more and nothing less.

### CODE REVIEW

**Comment by jesperpedersen** [comment]
I'm going to leave this pull request open until the full tablespace support is in

**Comment by Jubilee101** [comment]
@jesperpedersen Another problem, we probably need to add another layer of directory to wrap base directory and other tablespaces. Because now we need to tar them all together. If we just tar everything under their root directory, it will cause two problems: First tar functionality will try to tar itself since it's also under this directory. Secondly if there are multiple backups waiting to be archived, tar functionality will tar them all. I propose wrap it under a directory called `archive-timestamp/`

**Comment by Jubilee101** [comment]
Changes to accomodate tablespaces have been submitted. I had to make a lot changes and decisions to make this work:
1. Support relative path when restoring backup. Because creating symlink only takes absolute path, I have to convert relative path using realpath()
2. When restoring before making an archive, base directory and other tablespaces are wrapped under a directory called `archive-server_name-id`. And the tar directory is like this:
.                                                                                       .
= archive-server_name-id                                              = server_name-id 
    == server_name-id                               instead of      = server_name-id-tablespace1
    == server_name-id-tablespace1                                = server_name-id-tablespace2
    == server_name-id-tablespace2
This extra layer will ensure that the symlinks are valid when untared.
Restoring without making an archive is not affected. There won't be this extra archive layer.
4. There is a nasty bug in `pgmoneta_delete_directory`. Since we now support symlinks, using `stat` will cause it to trace the link. And error occurs if the target directory is deleted before it. Leaving the rest of the directory not deleted. The solution is simple however, replacing with `lstat` solves the problem.

File: src/libpgmoneta/archive.c [hunk 1]
```diff
@@ -39,27 +39,35 @@
 
 #include <archive.h>
 #include <archive_entry.h>
+#include <dirent.h>
 #include <stdio.h>
 
+void write_tar_file(struct archive* a, char* current_real_path, char* current_save_path);
```
Comment (jesperpedersen) [inline src/libpgmoneta/archive.c:45]: static

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by Jubilee101** [comment]
> The tablespace link in the archive doesn't work - it assume a path once extracted.
> 
> Create the archive in one location, move it to another and extract -> link is wrong

It is correct if you don't move it... The link is created during restoring. So it doesn't know where you are going to move and extract the archive to. We have little control on the extraction process. Unless we provide our own extraction functionality in the cli. Then we can change the symlink when extracting it. Otherwise we can only assume user makes the archive to the places they want in the first place.

**Comment by jesperpedersen** [comment]
Could you make the symlink relative ? Like, `13898 -> ../../primary-xyz-ts1/` ?

**Comment by Jubilee101** [comment]
> Could you make the symlink relative ? Like, `13898 -> ../../primary-xyz-ts1/` ?

Unfortunately no. Symlinks have to use absolute path..

**Comment by jesperpedersen** [comment]
I changed the API a bit... maybe `man 2 symlink` could work...

**Comment by Jubilee101** [comment]
I'll be damned... Took me quite some time to get right but it actually worked! I feel so dumb just arbitrarily saying no based on my limited experience.

**Comment by Jubilee101** [comment]
The link is like this: `16397 -> ../../primary-20230717193201-table_space_2/`
Tricky part is that you still have to use absolute path for the symlink location when creating symlink.

**Comment by Jubilee101** [comment]
> I changed the API a bit... maybe man 2 symlink could work...

Oh crap I didn't see this. I created a new util api using symlinkat. It does the trick

**Comment by Jubilee101** [comment]
I did some further verification. Looks like postgres takes symlinks using relative path. I can query tablespaces or create db under existing tablespaces normally. And moving archive around doesn't affect things, as long as the file structures under archive remain unchanged.

**Comment by jesperpedersen** [comment]
Rebased, and merged.

Thanks for your contribution !

**Pull request closed without merging**